src/libpmemfile-posix/block_array.c

```diff
@@ -0,0 +1,273 @@
+	/*
+	 * ... or if there is more than one block_array, it is
+	 * the one linked to it with the next field.
+	 */
+	if (!TOID_IS_NULL(binfo->arr->next))
+		binfo->arr = D_RW(binfo->arr->next);
+
+	binfo->idx = 0;
+
+	/* Find the first free entry in the block array. */
+	while (binfo->idx < binfo->arr->length &&
+	    binfo->arr->blocks[binfo->idx].size != 0)
+		binfo->idx++;
+}
+
+static bool
+has_free_block_entry(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	return binfo->idx < binfo->arr->length;
+}
+
+static void
+allocate_new_block_array(struct pmemfile_vinode *vinode)
+{
+	TOID(struct pmemfile_block_array) new =
+			TX_ZALLOC(struct pmemfile_block_array, FILE_PAGE_SIZE);
+	D_RW(new)->length = (uint32_t)
+			((page_rounddown(pmemobj_alloc_usable_size(new.oid)) -
+			sizeof(struct pmemfile_block_array)) /
+			sizeof(struct pmemfile_block));
+
+	D_RW(new)->next = vinode->inode->file_data.blocks.next;
+	TX_SET_DIRECT(&vinode->inode->file_data.blocks, next, new);
+	vinode->first_free_block.arr = D_RW(new);
+	vinode->first_free_block.idx = 0;
+}
+
+/*
+ * is_zeroed -- check if given memory range is all zero
+ */
+static inline bool
+is_zeroed(const void *addr, size_t len)
+{
+	/* XXX optimize */
+	const char *a = (const char *)addr;
+	while (len-- > 0)
+		if (*a++)
+			return false;
+	return true;
+}
+
+static struct pmemfile_block *
+acquire_new_entry(struct pmemfile_vinode *vinode)
+{
+	ASSERT(has_free_block_entry(vinode));
+
+	struct block_info *binfo = &vinode->first_free_block;
+	struct pmemfile_block *block = binfo->arr->blocks + binfo->idx++;
+
+	ASSERT(is_zeroed(block, sizeof(*block)));
+
+	/* XXX, snapshot separated to let pmemobj use small object cache  */
+	pmemobj_tx_add_range_direct(block, 32);
+	pmemobj_tx_add_range_direct((char *)block + 32, 32);
+	COMPILE_ERROR_ON(sizeof(*block) != 64);
+
+	return block;
+}
+
+struct pmemfile_block *
+block_list_insert_after(struct pmemfile_vinode *vinode,
+			struct pmemfile_block *prev)
+{
+	update_first_block_info(vinode);
+
+	if (!has_free_block_entry(vinode))
+		allocate_new_block_array(vinode);
+
+	struct pmemfile_block *block = acquire_new_entry(vinode);
+
+	if (prev == NULL) {
+		if (vinode->first_block != NULL) {
+			block->next = blockp_as_oid(vinode->first_block);
+			TX_SET_DIRECT(vinode->first_block,
+			    prev, blockp_as_oid(block));
+		}
+		vinode->first_block = block;
+	} else {
+		block->prev = blockp_as_oid(prev);
+		block->next = prev->next;
+		TX_SET_DIRECT(prev, next, blockp_as_oid(block));
+		struct pmemfile_block *next = D_RW(block->next);
+		if (next != NULL)
+			TX_SET_DIRECT(next, prev, blockp_as_oid(block));
+	}
+
+	return block;
+}
+
+static struct pmemfile_block *
+last_used_block(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	ASSERT(binfo->idx > 0);
+
+	return binfo->arr->blocks + (binfo->idx - 1);
+}
+
+static void
+unlink_block(struct pmemfile_block *block)
```

Comment:
Oh, I see, we forgot to add some comments about how `vinode->first_block` is handled